File: 039/combination-sum.go
package main

import (
	"fmt"
	"sort"
)

func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)

	var res [][]int
	var solution []int

	combinationSumHelper(candidates, solution, target, &res)

	return res
}

func combinationSumHelper(candidates []int, solution []int, target int, result *[][]int) {
	if target == 0 {
		*result = append(*result, solution)
	}

	if len(candidates) == 0 || target < candidates[0] {
		return
	}

	solution = solution[:len(solution):len(solution)]

	combinationSumHelper(candidates, append(solution, candidates[0]), target-candidates[0], result)
	combinationSumHelper(candidates[1:], solution, target, result)
}

func main() {
	candidates, target := []int{2, 3, 6, 7}, 7
	fmt.Println(combinationSum(candidates, target))
}

File: 039/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
function combinationSum(candidates, target) {
    candidates.sort((a, b) => a - b);
    let res = [];
    let out = [];

    function combinationSumHelper(candidates, target, start, out, res) {
        if (target < 0) return;
        else if (target == 0) res.push(out.map(i => i));
        else {
            for (let i = start; i < candidates.length; i++) {
                out.push(candidates[i]);
                combinationSumHelper(candidates, target - candidates[i], i, out, res);
                out.pop();
            }
        }
    };

    combinationSumHelper(candidates, target, 0, out, res);
    return res;
}

let candidates = [2, 3, 6, 7], target = 7;
console.log(combinationSum(candidates, target));